File: internal/infrastructure/clock/clock.go
package clock

import (
	"time"

	"github.com/memlore/memlore/internal/application/ports"
)

// SystemClock returns UTC wall-clock time.
type SystemClock struct{}

func (SystemClock) Now() time.Time {
	return time.Now().UTC()
}

// FixedClock is a test double with a constant instant.
type FixedClock struct {
	Instant time.Time
}

func (c FixedClock) Now() time.Time {
	return c.Instant.UTC()
}

var (
	_ ports.Clock = SystemClock{}
	_ ports.Clock = FixedClock{}
)

File: src/memlore/infrastructure/clock.py
from __future__ import annotations

from datetime import UTC, datetime

from memlore.application.ports.clock import Clock


class SystemClock:
    def now(self) -> datetime:
        return datetime.now(UTC)


class FixedClock:
    """Test double implementing Clock."""

    def __init__(self, instant: datetime) -> None:
        self._instant = instant

    def now(self) -> datetime:
        return self._instant


def as_clock(clock: SystemClock | FixedClock) -> Clock:
    return clock
